File: old/ivas-go/internal/domain/repository/question_repository.go
package repository

import (
	"github.com/ivas/internal/domain/entity"
	"gorm.io/gorm"
)

// QuestionRepository handles question database operations
type QuestionRepository struct {
	db *gorm.DB
}

// NewQuestionRepository creates a new question repository
func NewQuestionRepository(db *gorm.DB) *QuestionRepository {
	return &QuestionRepository{db: db}
}

// Create creates a new question
func (r *QuestionRepository) Create(question *entity.Question) error {
	return r.db.Create(question).Error
}

// CreateBatch creates multiple questions in a transaction
func (r *QuestionRepository) CreateBatch(questions []entity.Question) error {
	return r.db.Create(&questions).Error
}

// FindByID finds a question by ID
func (r *QuestionRepository) FindByID(id string) (*entity.Question, error) {
	var question entity.Question
	err := r.db.First(&question, "id = ?", id).Error
	if err != nil {
		return nil, err
	}
	return &question, nil
}

// FindByAssignmentID finds all questions for an assignment
func (r *QuestionRepository) FindByAssignmentID(assignmentID string) ([]entity.Question, error) {
	var questions []entity.Question
	err := r.db.Where("assignment_id = ?", assignmentID).Find(&questions).Error
	return questions, err
}

// FindByAssignmentIDWithFilters finds questions with optional filters
func (r *QuestionRepository) FindByAssignmentIDWithFilters(
	assignmentID string,
	status *string,
	competency *string,
	questionType *string,
) ([]entity.Question, error) {
	query := r.db.Where("assignment_id = ?", assignmentID)

	if status != nil && *status != "" {
		query = query.Where("status = ?", *status)
	}
	if competency != nil && *competency != "" {
		query = query.Where("competency = ?", *competency)
	}
	if questionType != nil && *questionType != "" {
		query = query.Where("question_type = ?", *questionType)
	}

	var questions []entity.Question
	err := query.Order("created_at DESC").Find(&questions).Error
	return questions, err
}

// Update updates a question
func (r *QuestionRepository) Update(question *entity.Question) error {
	return r.db.Save(question).Error
}

// UpdateStatus updates question status
func (r *QuestionRepository) UpdateStatus(id string, status entity.QuestionStatus) error {
	return r.db.Model(&entity.Question{}).Where("id = ?", id).Update("status", status).Error
}

// UpdateType updates question type
func (r *QuestionRepository) UpdateType(id string, qType entity.QuestionType) error {
	return r.db.Model(&entity.Question{}).Where("id = ?", id).Update("question_type", qType).Error
}

// SoftDelete archives a question (soft delete)
func (r *QuestionRepository) SoftDelete(id string) error {
	return r.UpdateStatus(id, entity.QuestionStatusArchived)
}

File: app/repositories/question_repository.py
"""
Question Repository

Data access layer for Question models. Provides CRUD operations for questions,
such as creating new generated/manual questions, retrieving questions by assignment,
updating question details, and archiving (soft-delete). These methods interact
directly with the database session and return SQLAlchemy ORM objects to the Services.
"""

from typing import Optional

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.question import Question, QuestionStatus, QuestionType


class QuestionRepository:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def create(self, question: Question) -> Question:
        self.session.add(question)
        await self.session.flush()
        await self.session.refresh(question)
        return question

    async def create_batch(self, questions: list[Question]) -> list[Question]:
        self.session.add_all(questions)
        await self.session.flush()
        return questions

    async def find_by_id(self, question_id: str) -> Optional[Question]:
        result = await self.session.execute(
            select(Question).where(Question.id == question_id)
        )
        return result.scalar_one_or_none()

    async def find_by_assignment_id(self, assignment_id: str) -> list[Question]:
        result = await self.session.execute(
            select(Question).where(Question.assignment_id == assignment_id)
        )
        return list(result.scalars().all())

    async def find_by_assignment_id_with_filters(
        self,
        assignment_id: str,
        status: Optional[str] = None,
        competency: Optional[str] = None,
        question_type: Optional[str] = None,
    ) -> list[Question]:
        stmt = select(Question).where(Question.assignment_id == assignment_id)

        if status:
            stmt = stmt.where(Question.status == status)
        if competency:
            stmt = stmt.where(Question.competency == competency)
        if question_type:
            stmt = stmt.where(Question.question_type == question_type)

        stmt = stmt.order_by(Question.created_at.desc())
        result = await self.session.execute(stmt)
        return list(result.scalars().all())

    async def update(self, question: Question) -> Question:
        await self.session.merge(question)
        await self.session.flush()
        return question

    async def update_status(self, question_id: str, status: QuestionStatus) -> None:
        await self.session.execute(
            update(Question)
            .where(Question.id == question_id)
            .values(status=status)
        )
        await self.session.flush()

    async def update_type(self, question_id: str, q_type: QuestionType) -> None:
        await self.session.execute(
            update(Question)
            .where(Question.id == question_id)
            .values(question_type=q_type)
        )
        await self.session.flush()

    async def soft_delete(self, question_id: str) -> None:
        await self.update_status(question_id, QuestionStatus.archived)
